File: clojure/math.clj
(defn div?
  "Returns true if a is evenly divisible by m"
  [a m]
  (zero? (rem a m)))

(defn psqr?
  [n]
  (let [x (Math/sqrt n)]
    (== (int x) x)))

(defn sum
  "Accepts a seq/coll and returns the sum of all elements in the
  seq/coll"
  [ls]
  (apply +' ls))

(defn product
  "Accepts a seq/coll and returns the product of all elements in the
  seq/coll"
  [ls]
  (apply *' ls))

(defn square [x] (* x x))
(defn cube [x] (* x x x))

(defn expt-big
  "Returns a to the m-th power, both must be zero or positive integers"
  [a m]
  (loop [i 0 res 1]
    (if (= i m)
      res
      (recur (inc i) (*' a res)))))

(defn sum-digits
  "Returns the sum of all digits in n"
  [n]
  (sum (map #(read-string (str %)) (str n))))

(defn expt
  [a m]
  (if (zero? m) 1 (* a (expt a (dec m)))))

(defn gcd
  "Accepts two numbers and returns the greatest common divisors of
  those numbers"
  [a b]
  (loop [i a j b]
    (if (= i j)
      i
      (if (> i j)
        (recur j (- i j))
        (recur i (- j i))))))

(defn prime?
  "Accepts a number and returns true if it is a prime, and false
  otherwise"
  [p]
  (cond (<= p 20) (if (some #(= % p) [2 3 5 7 11 13 17 19]) true false)
        (even? p) false
        :else (let [lim (inc (Math/sqrt p))]
                (loop [i 3]
                  (if (> i lim)
                    true
                    (if (zero? (rem p i))
                      false
                      (recur (+ 2 i))))))))

(defn rude-lcm
  [[a & xs] res]
  (if (empty? xs)
    (conj res a)
    (if (some #(zero? (rem % a)) xs)
      (-> #(if (zero? (rem % a)) (quot % a) %)
          (map xs)
          (rude-lcm (if (prime? a) (conj res a) res)))
      (rude-lcm xs (conj res a)))))

(defn lcm
  "Accepts a seq/coll of numbers and returns the least common multiple
  of all numbers in the seq/coll"
  [ls]
  (product (rude-lcm ls [])))

(defn factors
  "Accepts a number n and returns the factors of n"
  [n]
  (loop [i 1 res []]
    (if (>= (* i i) n)
      (if (= (* i i) n) (conj res i) res)
      (recur (inc i)
             (if (= 0 (rem n i))
               (if (= i (quot n i)) (conj res i) (conj res i (quot n i)))
               res)))))

(defn count-factors
  "Accepts a number n and returns the factors of n"
  [n]
  (loop [i 1 res 0]
    (if (>= (* i i) n)
      (if (= (* i i) n)
        (inc res)
        res)
      (recur (inc i)
             (if (= 0 (rem n i))
               (if (= i (quot n i))
                 (inc res)
                 (+ 2 res))
               res)))))

(defn count-divs
  "Return the number of divisors of n"
  [n]
  (if (= n 1)
    1
    (let [plim (inc (Math/sqrt n))]
      (if (even? n)
        (loop [i 2 lim (quot n 2) res 2]
          (if (or (> i plim) (>= i lim))
            (if (= 4 n)
              3
              res)
            (if (zero? (rem n i))
              (let [tmp (quot n i)]
                (if (= tmp i)
                  (inc res)
                  (recur (inc i)
                         tmp
                         (+ 2 res))))
              (recur (inc i)
                     lim
                     res))))
        (loop [i 3 lim (quot n 3) res 2]
          (if (or (> i plim) (>= i lim))
            res
            (if (zero? (rem n i))
              (let [tmp (quot n i)]
                (if (= tmp i)
                  (inc res)
                  (recur (+ 2 i)
                         tmp
                         (+ 2 res))))
              (recur (+ 2 i)
                     lim
                     res))))))))


(defn next-prime
  "Returns the smallest prime that is larger than x"
  [x]
  (cond (= 1 x) 2
        (= 2 x) 3
        (even? x) (next-prime (inc x))
        :else (loop [i (+ 2 x)]
                (if (prime? i) i (recur (+ 2 i))))))


;; (defn sum-primes
;;   "Returns the smallest prime that is larger than x"
;;   [lim]
;;   (+ 2 (loop [i (+ 2 x)]
;;          (if (prime? i) i (recur (+ 2 i))))))

(defn prime-list
  "Returns the n first positive prime numbers"
  [n]
  (loop [i 1 cur 2 res '()]
    (if (= i n)
      (cons cur res)
      (recur (inc i)
             (next-prime cur)
             (cons cur res)))))

(defn prime-list-vector
  "Returns the n first positive prime numbers"
  [n]
  (loop [i 1 cur 2 res []]
    (if (= i n)
      (conj res cur)
      (recur (inc i)
             (next-prime cur)
             (conj res cur)))))

(defn primes-under
  "Returns all positive primes less than n"
  [n]
  (loop [i 2 res []]
    (if (> i n)
      res
      (recur (next-prime i) (conj res i)))))

(defn suma-prima
  "Returns the sum of n first positive prime numbers"
  [n ctype]
  (if (= :list ctype)
    (reduce + (prime-list n))
    (reduce + (prime-list-vector n))))

(defn abs
  [n]
  (if (neg? n) (- n) n))

(defn now
  []
  (java.util.Date.))

(defn numcol
  "Returns a vector of digits of a number n"
  [n]
  (loop [i n res '()]
    (if (< i 10)
      (cons i res)
      (recur (quot i 10) (cons (rem i 10) res)))))


(defn colnum
  "Returns a number made from a vector of digits"
  [ls]
  (loop [[x & xs] ls res 0]
    (if (empty? xs)
      (+ (* 10 res) x)
      (recur xs (+ (* 10 res) x)))))

(defn palin?
  "Returns true if n is a palindromic number"
  [n]
  (let [col (numcol n)]
    (= col (reverse col))))

(defn palins [n]
  (let [gen-pal (fn gen-pal [n mcol]
                  (let [nn (inc n)
                        ncol (numcol nn)
                        stat (if (odd? (count mcol))
                               (not= (* 2 (count ncol))
                                     (inc (count mcol)))
                               (not= (* 2 (count ncol))
                                     (count mcol)))]
                    (if stat
                      (colnum (concat ncol
                                      (if (odd? (count mcol))
                                        (reverse (butlast (butlast ncol)))
                                        (reverse (butlast ncol)))))
                      (colnum (concat ncol
                                      (if (odd? (count mcol))
                                        (reverse (butlast ncol))
                                        (reverse ncol)))))))
        npal (fn [n]
               (let [mcol (numcol n)
                     lcol (vec (take (quot (count mcol) 2) mcol))
                     rlcol (reverse lcol)
                     rcol (take-last (quot (count mcol) 2) mcol)
                     midnum (first (drop (quot (count mcol) 2) mcol))
                     stat (odd? (count mcol))]
                 (if (<= (count mcol) 2)
                   (loop [i (inc n)]
                     (if (palin? i)
                       i
                       (recur (inc i))))
                   (if (palin? n)
                     (gen-pal (colnum (if stat
                                        (conj lcol midnum)
                                        lcol)) mcol)
                     (if (> (colnum rlcol)
                            (colnum rcol))
                       (colnum (concat lcol
                                       (if stat
                                         [midnum]
                                         [])
                                       rlcol))
                       (gen-pal (colnum (if stat
                                          (conj lcol midnum)
                                          lcol)) mcol))))))
        m (if (palin? n)
            n
            (npal n))]
    (iterate npal m)))

(defn palins-under
  "Returns all palindromic numbers less than n"
  [n]
  (take-while #(< % n) (palins 1)))


(defn pfactors
  [nt]
  (let [lim (Math/sqrt nt)]
    (loop [p 2 n nt res []]
      (let [d (quot n p) r (rem n p)]
        (if (> p lim)
          (sort (distinct (conj res p d)))
          (if (zero? r)
            (if (or (= 1 d) (prime? d))
              (sort (distinct (conj res p d)))
              (recur 2 d (conj res p)))
            (recur (next-prime p) n res)))))))

(defn totient
  [n]
  (let [pfacts (pfactors n)]
    (* n (product (map #(- 1 (/ 1 %)) pfacts)))))

File: haskell/math.hs
module Math where

import Data.List

primeHelper :: Int -> Int -> Bool
primeHelper p i
  | (i*i) > p = True
  | 0 == (rem p i) = False
  | otherwise = primeHelper p (i + 2)

-- it returns true if p is prime and false otherwise
prime :: Int -> Bool
prime p
  | p <= 10 = elem p [2,3,5,7]
  | even p = False
  | otherwise = primeHelper p 3

factorsHelper :: Int -> Int -> [Int] -> [Int]
factorsHelper n i res
  | i * i > n = res
  | 0 == rem n i = factorsHelper n (succ i) newRes
  | otherwise = factorsHelper n (succ i) res
  where newRes = if i == (quot n i) then i:res else i:(quot n i):res

-- it returns the list of integer factors of n
factors :: Int -> [Int]
factors 1 = [1]
factors 2 = [1,2]
factors 3 = [1,3]
factors 6 = [1,2,3,6]
factors 12 = [1,2,3,4,6,12]
factors n = factorsHelper n 1 []


-- it returns true if p is a palindrom
isPalin :: Int -> Bool
isPalin p = (show p) == (reverse $ show p)

-- it returns the list of prime factors which multiplied to produce
-- lcm of all numbers in the list
rudeLCM :: [Int] -> [Int] -> [Int]
rudeLCM (a:xs) res
  | null xs = a:res
  | any (\x -> 0 == (rem x a)) xs = rudeLCM newXs newRes
  | otherwise = rudeLCM xs (a:res)
  where newXs = map (\x -> (if (0 == (rem x a)) then (quot x a) else x)) xs
        newRes = if (prime a) then a:res else res

-- it returns the first positive prime number greater than x
nextPrime :: Int -> Int
nextPrime x
  | x == 2 = 3
  | even x = if prime $ succ x then succ x else nextPrime $ succ x
  | otherwise = if prime $ 2 + x then 2 + x else nextPrime $ 2 + x

primeListHelper :: Int -> Int -> Int -> [Int] -> [Int]
primeListHelper n i cur res
  | n == i = cur:res
  | otherwise = primeListHelper n (succ i) (nextPrime cur) (cur:res)

-- it returns n first positive prime numbers
primeList :: Int -> [Int]
primeList n = primeListHelper n 1 2 []


sumaPrimaHelper :: Int -> Int -> Int -> Int -> Int
sumaPrimaHelper n i cur res
  | n == i = cur + res
  | otherwise = sumaPrimaHelper n (succ i) (nextPrime cur) (cur+res)

-- it returns the sum of n first positive prime numbers
sumaPrima :: Int -> Int
sumaPrima n = sumaPrimaHelper n 1 2 0

primesUnder :: Int -> [Int]
primesUnder n = takeWhile (< n) $ iterate nextPrime 2 

sumPrimesHelper :: Int -> Int -> Int -> Int
sumPrimesHelper n i res
  | i > n = res
  | otherwise = sumPrimesHelper n (nextPrime i) (i + res)

-- Returns true if n is a perfect square

psquare :: Int -> Bool
psquare n = ceiling x == floor x
  where x = sqrt.fromIntegral $ n

-- it returns the sum of all primes less than n
sumPrimes :: Int -> Int
sumPrimes n = sumPrimesHelper n 2 0

-- it returns the list of digits of a number n
numcol :: Integer -> [Int]
numcol n = if n < 10
           then [fromInteger n]
           else (numcol $ quot n 10) ++ [rem (fromInteger n) 10]

-- it returns a number from a list of digits
colnum :: [Int] -> Integer
colnum ls
  | null ls = 0
  | otherwise = (10 * (toInteger (colnum $ init ls))) + (toInteger $ last ls)

-- yeah... just to simplify things
div' a b = (0 == rem a b)

-- it returns a list of primes less than lim
sieve lim = takeWhile (< lim) (2 : primes [3,5..])

-- the helper for sieve
primes (x:xs) = x : deleteBy (\x n -> div' n x) x (primes xs)

distinct lst =  map head.group.sort $ lst








               


